media_utils/video_puppeteer.py

```
@@ -150,49 +138,29 @@ class VideoException(Exception):
-
-# One simple hack: There should be two active readers at any one time. If
-# there are not exactly two, something is horribly wrong.
-
 
 def playback_done(video):
-    active_re = re.compile('active=true', re.DOTALL)
-
     # If we are near the end and there is still a video element, then
-    # we are essentially done. Any further progress might set on of
-    # the stream active state to false, and we will raise in that
-    # case inappropriately.
+    # we are essentially done. If this happens to be last time we are polled
+    # before the video ends, we won't get another chance.
 
-    if abs(video.remaining_time) < 2.0:
+    remaining_time = video.remaining_time
+    if abs(remaining_time) < video.interval:
         return True
 
-    # Now, parse out playing info. If there is no debug info, then the
-    # video element has disappeared somehow, and that' a problem.
-    # Otherwise, there should be one active video and one active audio
-    # stream. (TODO: what if there are multiple tabs with video content?)
-    # If not, there is a problem. Past this point, the video is
-    # either still playing or there is a problem, so this will never
-    # return True.
-
-    debug_lines = video.get_debug_lines()
-    if debug_lines is not None:
-        num_active = len(active_re.findall(' '.join(debug_lines)))
-        if num_active == 2:
-            return False
-
-    raise VideoException('Did not find exactly one audio and one video '
-                         'active reader - %s'
-                         % video)
+    # Check to see if the video has stalled. Accumulate the amount of lag
+    # since the video started, and if it is too high, then raise.
+    elapsed_current_time = video.current_time - video.start_time
+    elapsed_wall_time = clock() - video.start_wall_time
+    lag = elapsed_wall_time - elapsed_current_time
+    if lag > video.stall_wait_time:
+        raise VideoException('Video %s stalled' % video)
```

If the video is not stalled right at this moment, but lagged more than `stall_wait_time` in the past, do we still want to raise an exception? I'm leaning toward 'no': if playback is progressing right now, it means it has recovered from previous pauses (which may be due to buffering and bandwidth), and it's not necessarily a problem with MSE, especially if `stall_wait_time` is small, like 3 or 10s. 

If there is a lot of delay due to buffering, that could be a problem on our side too, but we're going to be measuring that specifically in the bandwidth-limiting tests so we shouldn't worry about it here. Also, if there is a large lag during this test, the `Wait` used in `run_playback` will raise a `TimeoutException`.

__Summary__: I'd do something like `if lag > video.stall_wait_time and self.measure_progress < 0.1:` raise. 

Eventually we could also refine stall detection by looking at `_debug_data`: if we see a lag and `NextGetSampleIndex` is 0, something is definitely broken.